### PR TITLE
feat: table filterSearch supports function

### PR DESCRIPTION
**Pull request by heiyu4585**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [x] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [x] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 🔗 Related issue link

https://github.com/ant-design/ant-design/issues/33935
### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |    feat: table filterSearch supports function `(input, record) => record.value.indexOf(input) > -1`    |
| 🇨🇳 Chinese |    功能: 表格`filterSearch`属性支持方法`(input, record) => record.value.indexOf(input) > -1`    |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

File: components/table/hooks/useFilter/FilterDropdown.tsx [hunk 1]
```diff
@@ -86,7 +92,7 @@ export interface FilterDropdownProps<RecordType> {
   filterState?: FilterState<RecordType>;
   filterMultiple: boolean;
   filterMode?: 'menu' | 'tree';
-  filterSearch?: boolean;
+  filterSearch?: boolean | Function;
```
Comment (MadCcc) [inline components/table/hooks/useFilter/FilterDropdown.tsx:95]: 这里的 Function 定义明确一些？

File: components/table/index.zh-CN.md [hunk 2]
```diff
@@ -137,7 +137,7 @@ const columns = [
 | filterIcon | 自定义 filter 图标。 | ReactNode \| (filtered: boolean) => ReactNode | false |  |
 | filterMultiple | 是否多选 | boolean | true |  |
 | filterMode | 指定筛选菜单的用户界面 | 'menu' \| 'tree' | 'menu' | 4.17.0 |
-| filterSearch | 筛选菜单项是否可搜索 | Boolean | false | 4.17.0 |
+| filterSearch | 筛选菜单项是否可搜索 | Boolean \| function(input, record){} | false | 4.19.0 |
```
Comment (yoyo837) [inline components/table/index.zh-CN.md:140]: since 直接改到4.19.0 是不是不太好，保留一下 boolean的since?

Comment (heiyu4585) [reply components/table/index.zh-CN.md:140]: 没太理解,文档需要修改 还是什么

File: components/table/hooks/useFilter/FilterDropdown.tsx [hunk 3]
```diff
@@ -42,7 +42,7 @@ function renderFilterItems({
   filteredKeys: Key[];
   filterMultiple: boolean;
   searchValue: string;
-  filterSearch: Function | boolean;
+  filterSearch: ((input: string, record: {}) => boolean) | boolean;
```
Comment (yoyo837) [inline components/table/hooks/useFilter/FilterDropdown.tsx:45]: 先boolean还是先函数，写成一致的吧

Comment (heiyu4585) [reply components/table/hooks/useFilter/FilterDropdown.tsx]: ok

File: components/table/hooks/useFilter/FilterDropdown.tsx (hunk 1 above)
Comment (heiyu4585) [reply components/table/hooks/useFilter/FilterDropdown.tsx]: ok

File: components/table/index.zh-CN.md (hunk 2 above)
Comment (yoyo837) [reply components/table/index.zh-CN.md]: filterSearch : boolean注明4.17.0添加，filterSearch : function 注明 4.19.0添加。

Comment (heiyu4585) [reply components/table/index.zh-CN.md]: ok

File: components/table/hooks/useFilter/FilterSearch.tsx [hunk 4]
```diff
@@ -6,7 +6,7 @@ import { TableLocale } from '../../interface';
 interface FilterSearchProps {
   value: string;
   onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
-  filterSearch: Boolean;
+  filterSearch: boolean | ((input: string, record: {}) => boolean);
```
Comment (MadCcc) [inline components/table/hooks/useFilter/FilterSearch.tsx:9]: 感觉抽个 type 出来好点

Comment (heiyu4585) [reply components/table/hooks/useFilter/FilterSearch.tsx:9]: ok

File: components/table/demo/filter-in-tree.md [hunk 5]
```diff
@@ -83,7 +83,7 @@ const columns = [
       },
     ],
     onFilter: (value, record) => record.address.startsWith(value),
-    filterSearch: true,
```
Comment (yoyo837) [inline components/table/demo/filter-in-tree.md:86]: @MadCcc `true`的demo留不留？

Comment (MadCcc) [reply components/table/demo/filter-in-tree.md:86]: 感觉可以留一下，这个 demo 是树形筛选的 demo 吧，可以新增一个 自定义筛选搜索条件的 demo @heiyu4585

Comment (heiyu4585) [reply components/table/demo/filter-in-tree.md]: ok

File: components/table/demo/filter-search.md [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+order: 6.2
+version: 4.19.0
+title:
+  en-US: Filter search
+  zh-CN: 自定义筛选
+---
+
+## zh-CN
+
+`filterSearch` 用于开启筛选项的搜索,通过`filterSearch:(input, record) => boolean`设置自定义筛选方法
```
Comment (MadCcc) [inline components/table/demo/filter-search.md:11]:
```suggestion
`filterSearch` 用于开启筛选项的搜索，通过 `filterSearch:(input, record) => boolean` 设置自定义搜索方法
```

File: components/table/demo/filter-search.md [hunk 7]
```diff
@@ -0,0 +1,100 @@
+---
+order: 6.2
+version: 4.19.0
+title:
+  en-US: Filter search
+  zh-CN: 自定义筛选
```
Comment (MadCcc) [inline components/table/demo/filter-search.md:6]:
```suggestion
  zh-CN: 自定义筛选的搜索
```

File: components/table/index.zh-CN.md [hunk 8]
```diff
@@ -137,7 +137,7 @@ const columns = [
 | filterIcon | 自定义 filter 图标。 | ReactNode \| (filtered: boolean) => ReactNode | false |  |
 | filterMultiple | 是否多选 | boolean | true |  |
 | filterMode | 指定筛选菜单的用户界面 | 'menu' \| 'tree' | 'menu' | 4.17.0 |
-| filterSearch | 筛选菜单项是否可搜索 | Boolean | false | 4.17.0 |
+| filterSearch | 筛选菜单项是否可搜索 | Boolean \| function(input, record):boolean | false | Boolean:4.17.0 function:4.19.0 |
```
Comment (MadCcc) [inline components/table/index.zh-CN.md:140]:
```suggestion
| filterSearch | 筛选菜单项是否可搜索 | boolean \| function(input, record):boolean | false | Boolean:4.17.0 function:4.19.0 |
```

File: components/table/index.en-US.md [hunk 9]
```diff
@@ -130,7 +130,7 @@ One of the Table `columns` prop for describing the table's columns, Column has t
 | filterIcon | Customized filter icon | ReactNode \| (filtered: boolean) => ReactNode | - |  |
 | filterMultiple | Whether multiple filters can be selected | boolean | true |  |
 | filterMode | To specify the filter interface | 'menu' \| 'tree' | 'menu' | 4.17.0 |
-| filterSearch | Whether to be searchable for filter menu | Boolean | false | 4.17.0 |
+| filterSearch | Whether to be searchable for filter menu | Boolean \| function(input, record):boolean | false | Boolean:4.17.0 function:4.19.0 |
```
Comment (MadCcc) [inline components/table/index.en-US.md:133]:
```suggestion
| filterSearch | Whether to be searchable for filter menu | boolean \| function(input, record):boolean | false | Boolean:4.17.0 function:4.19.0 |
```

Comment (heiyu4585) [reply components/table/index.en-US.md]: ok

**Review by MadCcc — APPROVED** [review]

**Comment by MadCcc** [comment]
怎么关了呀

**Comment by yoyo837** [comment]
> 怎么关了呀

我正想问

**Comment by heiyu4585** [comment]
> > 怎么关了呀
> 
> 我正想问

sorry,提了个新pr,怕污染,我需要怎么操作下

**Pull request closed without merging**